File: core/src/allowlist-policy.mjs
import { isHomeUrl, rawHost } from './url-policy.mjs';

export function isHostAllowed(hosts, host) {
  return host.length > 0 && hosts.includes(host);
}

export function setHostAllowed(hosts, host, allowed) {
  if (host.length === 0) {
    return hosts.slice();
  }
  const without = hosts.filter((item) => item !== host);
  return allowed ? without.concat([host]) : without;
}

export function adsBlockEnabledForUrl(url, hosts, blockAds) {
  if (!blockAds || isHomeUrl(url)) {
    return false;
  }
  return !isHostAllowed(hosts, rawHost(url));
}

File: ios/ZhuoBrowser/AllowListPolicy.swift
import Foundation

enum AllowListPolicy {
    static func isHostAllowed(_ hosts: [String], _ host: String) -> Bool {
        !host.isEmpty && hosts.contains(host)
    }

    static func setHostAllowed(_ hosts: [String], host: String, allowed: Bool) -> [String] {
        if host.isEmpty {
            return hosts
        }
        let without = hosts.filter { $0 != host }
        return allowed ? without + [host] : without
    }

    static func adsBlockEnabled(for url: String, hosts: [String], blockAds: Bool) -> Bool {
        if !blockAds || URLPolicy.isHomeURL(url) {
            return false
        }
        return !isHostAllowed(hosts, URLPolicy.rawHost(url))
    }
}
